### ISSUE TITLE
Introduce initialSimplified() operator to support nonlinear model inversion (was: Extended homotopy() operator to support nonlinear model inversion)

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#11272]
When modelling power generation systems (but also other kind of systems), the following use scenario is usually confronted:
- the plant is modelled by a nonlinear dynamical system with top-level inputs (actuator signals) and outputs (sensor signals)
- one wants to initialize the system in steady state, which leads to large nonlinear systems of algebraic equations that are potentially difficult to solve
- when doing _forward initialization_, the inputs are fixed to some values, and the corresponding outputs are computed
- when doing _backward initialization_, the output are fixed to some target values, and the corresponding inputs are computed backwards
- when initializing the system in nominal/design conditions, one usually knows the results of some design computations carried out with other tools, which gives a pair of input/output values; however, the Modelica model is usually slightly different from the one used in the design tool, so when the design values of the inputs are given to the system, the computed outputs slightly differ from the design ones; the goal of backward initialization is to set the outputs to the exact design values, and to compute the corresponding exact values of the inputs that will lead to those outputs _with the Modelica model_
- in many cases, possibly also thanks to clever use of homotopy, forward initialization is numerically less challenging than backwards initialization; solving backward initialization is often much harder, and sometimes practically impossible (i.e. one tries and tries, and eventually gives up)

It would be really nice that after struggling for a significant amount of time (possibly months, ask @matteodepascali about his PhD project 😅) to get the forward initialization problem to converge, one could get convergence on the backward initialization out of the box, instead of continuing the struggle for a further unknown (possibly unlimited) amount of time to achieve that goal.

It turns out, this goal could be achieved by a simple extension of the homotopy operator. The idea is to add a third (optional) argument to the operator: `homotopy(actual, simplified, simplified_guess = simplified)`. 

The current semantics of `homotopy(actual, simplified)` is the following:
- for dynamic simulation (_regular_ equations), `homotopy(actual, simplified)` is replaced by `actual`.
- for the `lambda = 0` solution of the initialization problem (_initial-lambda0_ equations), it is replaced by `simplified`, exploiting all structural simplifications made possible by the simplified expression
- for all other values of `lambda` during the homotopy transformation of the _initial_ equations, it is replaced by `lambda*actual + (1 - lambda)*simplified`

The semantics of the extended `homotopy(actual, simplified, simplified_guess = simplified)` is the following
- for dynamic simulation (the _regular_ equations), `homotopy(actual, simplified, simplified_guess)` is replaced by `actual`.
- for the `lambda = 0` solution of the initialization problem, (the _initial-lambda0_ equations), a first guess is obtained by replacing `homotopy(actual, simplified, simplified_guess)` with `simplified_guess`, exploiting all structural simplifications made possible by the simplified expression. `simplified_guess` defaults to `simplified` if not explicitly supplied.
- for the solution of the _initial_ equations during the homotopy transformation, `homotopy(actual, simplified, simplified_guess)` is replaced by `lambda*actual + (1 - lambda)*simplified`. The solution for `lambda = 0` at the start of the homotopy transformation is obtained using the one found at the previous step as initial guess.

Note that the simplified expression needs to be chosen carefully to be not too "structurally different" from the actual one, otherwise it is likely that a singularity occurs for some value of lambda. The structure of the problem obtained with the simplified_guess expression can instead be completely different, as it won't be involved in the homotopy trasformation - only its numerical values matter.

I understand the implementation of this extension is trivial, as we already handle the _initial-lambda0_ equations separately from the `initial` ones.

The availability of this extended operator would solve the problem of robust backward initialization in a trivial way: if the plant model has input u and output y, one just needs to write:
```modelica
model Backward
  Plant plant;
  parameter Real u_des, y_des;
equation
  homotopy(y - y_des, y - y_des, u - u_des) = 0;
end Backward;
```
In this way, during the initial-lambda0 phase the equation `u - u_des = 0` is active, which means, the _forward_ initialization problem is solved. This computes approximate values for all the variables of the problem. Afterwards, the backward initialization problem starts being solved at `lambda = 0`, but already has good initial guesses, so it should not have convergence issues.

The following MWE demonstrates the usefulness of this extension:
```modelica
package TestInverse
  block B
    parameter Integer N = 2;
    parameter Real u_des = 50;
    parameter Real y_des = 5;
    Real x[N];
    Real y[N];
    Real u[N];
    Modelica.Blocks.Interfaces.RealInput u_in;
    Modelica.Blocks.Interfaces.RealOutput y_out;
  equation
    for i in 1:N loop
      x[i] = u[i] + 1e-6*sin(y[i]*u[i]);
      y[i] = log(x[i] + 1 + 1e-6*sin(y[i]));
    end for;
    u[1] = u_in;
    y_out = y[N]*u[1];
    u[2:N] = y[1:N - 1];
  annotation(
      experiment(StartTime = 0, StopTime = 1, Tolerance = 1e-6, Interval = 0.002));
end B;

  model TestForward
    B b(N = 20);
  equation
    b.u_in = b.u_des;
    annotation(
      experiment(StartTime = 0, StopTime = 0.11, Interval = 0.00022, Tolerance = 1e-6));
  end TestForward;

  model TestBackward
    B b(N = 20);
  equation
    b.y_out = b.y_des;
  annotation(
      experiment(StartTime = 0, StopTime = 1, Interval = 0.1));
 end TestBackward;

  model TestBackwardHomotopy
    B b(N = 20);
  equation
    homotopy(b.y_out - b.y_des, b.y_out - b.y_des, b.u_in - b.u_des) = 0;
  annotation(
      experiment(StartTime = 0, StopTime = 1, Interval = 0.1));
  end TestBackwardHomotopy;
end TestInverse;
```
Block `B` contains an algebraic system of equations, which represents the steady-state initialization equations of a nonlinear dynamical system. 

`TestForward` initializes in forward mode, i.e., the input is fixed to the design value, and the solution is successful. It gives an output value `b.y_out = 5.4364`, which is slightly different from the expected design value `b.y_des =  5`, but good enough as an initial guess.

`TestBackward` initializes in backward mode, i.e., the output is fixed to the design value `b.y_des = 5`, but the solver fails, because the backward problem is much harder and the Newton solver fails if proper initial guesses are not given. 

`TestBackwardHomotopy` uses the proposed approach to solve the forward problem first, leading to an output value `b.y_out =  5.4364`, which is close to the the design value `b.y_des = 5`. This (hopefully) provides good enough initial guesses to solve the backward initialization problem with `b.y_out = 5`.

**Comment by casella** [comment]
In the meantime I collected another interesting use case for this feature, on which I'm currently working with @CJLuiten.

 Long story short: he wants to use (static) Modelica models for design of fluid system components, and then re-use them for off-design computations, which requires to switch causality, while keeping the same equations. In principle, this can be done by keeping the same code, and relying on the Modelica tool to solve it differently. In practice, getting convergence on off-design problems with the same equations you wrote for the design problem is often challenging.

The solution is to use the proposed extension: first you solve the design problem, then you use it as an initial guess for the on-design simulation problem (which has the same solution, so it converges naturally), and then you gradually bring it to off-design conditions by homotopy, which will work in almost all cases, provided there exists a continuous path to those conditions.

So, this feature would actually become an key enabling factor to _actually make good use_ of the a-causal features of Modelica, without Damocles' sword of convergence failures.

**Comment by casella** [comment]
One important point is how to manage the transient until (hopefully) this extension is accepted into the language. 

Currently, the MLS [defines](https://specification.modelica.org/master/operators-and-expressions.html#modelica:homotopy) `homotopy(actual, simplified)` as a special purpose operator with function syntax, with two arguments.

I would like to experiment with this extended homotopy operator in OpenModelica, but I would try to avoid making the code illegal, so it can't be run (or maybe even parsed) in other Modelica tools. Ideally, I would like to write homotopy(actual, simplified, simplified_guess), which is interpreted as explained above by OpenModelica, while other tools just ignore the third argument and do regular homotopy.

I would handle this by adding this function definition in the models that use this extended homotopy operator:

```modelica
function homotopy "Extended homotopy function stub"
  input Real actual;
  input Real simplified;
  input Real simplified_guess;
  output Real expr;
algorithm
  expr = homotopy(actual, simplified);
annotation(Inline = true, __OpenModelica_ignoreFunctionDefinition);
end homotopy;
```
so that OpenModelica ignores this function definition thanks to the annotation and uses the built-in experimental operator instead, while other tools would just use this definition to replace the experimental three-argument homotopy operators with regular two-argument homotopy operator via inlining.

@perost do you think this is viable?

**Comment by casella** [comment]
Regarding the implementation, I think this can be done very easily. The current workflow for `homotopy(actual, simplified)` is:
- we replace it with `simplified` and generate the `initial-lambda0` code
- we replace it with `lambda*actual + (1 - lambda)*simplified` and generate the initial code
- we first run the `initial-lambda0` code, to compute the solution at `lambda = 0`
- we then perform the homotopy transformation with the `initial` code, using the result of the previous lambda step as an initial guess for the current one

When the experimental three-argument `homotopy(actual, simplified, simplified_guess)` is used, we should just generate the initial-lambda0 code by replacing it with `simplified_guess`, instead of `simplified`. That's it. Everything else would stay the same.

@phannebohm, I think this can be implemented in no time. What do you think?

**Comment by perost** [comment]
> [@perost](https://github.com/perost) do you think this is viable?

It's either very easy or very hard depending on what you want to achieve. If you need it to be named `homotopy` then it becomes very complicated since it will collide with the already existing `homotopy` operator.

But if the name doesn't matter then I guess you can already achieve what you want by using the `__OpenModelica_builtin=true` annotation, which will make the frontend assume it's a builtin function and just pass it through.

**Comment by casella** [comment]
> But if the name doesn't matter then I guess you can already achieve what you want by using the __OpenModelica_builtin=true annotation, which will make the frontend assume it's a builtin function and just pass it through.

OK, so when I want to use this experimental new operator I should define it as

```modelica
function homotopyExtended "Extended experimental homotopy function"
  input Real actual "Actual expression used for the simulation phase";
  input Real simplified "Simplified expression for homotopy at lambda = 0";
  input Real simplified_guess = simplified "Simplified expression used to compute the guess values of the lambda = 0 problem";
  output Real expr;
algorithm
  expr := homotopy(actual, simplified);
annotation(Inline = true, __OpenModelica_builtin=true);
end homotopyExtended;
```
right? Then @phannebohm could implement this in the backend by using the `simplified_guess` expression for the initial_lambda0 code generation instead of the `simplified` expression.

I need to check with Dymola that it does the inlining early enough that the inlined homotopy operator is used as expected.

**Comment by casella** [comment]
@phannebohm, one important point, I will need to use `homotopyExtended()` also for non-Real (i.e. Boolean or enumeration) parameters, in this case the simplified and actual inputs should be the same. For example:

```modelica
model MWE
  Boolean forward "=true solve system forward, otherwise backward";
  parameter Real u_fix0 = 1;
  parameter Real y_fix0 = 1;
  parameter Real y_fix = 10;
  Real u, y1, y2;

  function homotopyExtended "Extended experimental homotopy function"
    input Real actual "Actual expression used for the simulation phase";
    input Real simplified "Simplified expression for homotopy at lambda = 0";
    input Real simplified_guess = simplified "Simplified expression used to compute the guess values of the lambda = 0 problem";
    output Real expr;
  algorithm
    expr := homotopy(actual, simplified);
  annotation(Inline = true, __OpenModelica_builtin=true);
  end homotopyExtended;

equation
  forward = homotopyExtended(false, false, true);
  if forward then
    u = u_fix0;
  else
    y2 = homotopy(y_fix, y_fix0);
  end if;
  y1 = u^3 + sin(u);
  y2 = sqrt(y1) + sin(1e-6*u);
end MWE;
```
The expected behaviour is the following:
- The `initial_lambda0` code sets `u` to `u_fix0` and then solves the other two equations as simple assignment
- The initial equations are set up with homotopy, setting the `y2` variable instead. At `lambda = 0`, `y2` is set to` y_fix0`, which is very close to the value `y2` gets when solving `initial_lambda0`, so the Newton solver converges without problems. Then, a homotopy transformation starts to bring the value of `y2` to th much larger `y_fix` value.

The whole process allows to eventually solve the implicit equations obtained by setting `y2 = y_fix` reliably and without the need of setting any start values.

**Comment by phannebohm** [comment]
> I will need to use `homotopyExtended()` also for non-Real (i.e. Boolean or enumeration) parameters,

This will not work well and cause many headaches. Instead I suggest a new operator similar to `initial()`
```modelica
function initialLambda0 "Experimental operator for homotopy initialization"
  output Boolean isInitialLambda0 = false;
annotation(Inline = true, __OpenModelica_builtin=true);
end initialLambda0;
```

**Comment by casella** [comment]
Here is the updated MWE

```modelica
package TestInversePlant
  function initialSimplified "Returns true when solving the initialization problem with the simplified expressions of homotopy() operators"
    output Boolean isInitialSimplified = false;
  annotation(Inline = true, __OpenModelica_builtin=true);
  end initialSimplified;

  model Plant
    connector RealInput = input Real;
    connector RealOutput = output Real;
  
    parameter Integer N = 5 "Number of finite volumes";
    parameter Real NTU = 4 "Nominal NTU";
    final parameter Real Gtot = NTU "Nominal total heat conductance";
    parameter Real cp = 1 "Specific heat capacity";
    parameter Real kf = 0.1 "Fluid friction coefficient";
    parameter Real Kt = 1 "Turbine flow coefficient";
    parameter Real Tnom = 700 "Nominal TIT";
    parameter Real Tsnom = 400 "Nominal flue gas outlet temperature";
    parameter Real Pnom = 223 "Nominal turbine power";
    parameter Real pnom = 1.1 "Nominal fluid inlet pressure";
    parameter Real TOP = 0.1 "Nominal turbine outlet pressure";
    parameter Real gamma = 1.2 "cp/cv";
  
    RealInput w "Fluid flow rate";
    RealInput wfg "Flue gas flow rate";
  
    Real T[N + 1] "Fluid temperatures";
    Real p[N + 1] "Fluid pressures";
    Real Tfg[N + 1] "Flue gas temperatures";
    Real G[N] "Volume conductivity";
    Real Q[N] "Thermal powers";
    Real TIT = T[N + 1] "Turbine inlet temperature";
  
    RealOutput P "Turbine power";
    RealOutput Ts = Tfg[N+1] "Outlet flue gas temperature";
    RealOutput TIP = p[N + 1] "Turbine inlet pressure";
    RealOutput pin = p[1] "Fluid inlet pressure";
  equation
    T[1] = 300;
    Tfg[1] = 800;
    for i in 1:N loop
      0 = w*cp*(T[i] - T[i + 1]) + Q[i];
      0 = wfg*cp*(Tfg[N - i + 1] - Tfg[N - i + 2]) - Q[i];
      Q[i] = ((Tfg[N - i + 2] + Tfg[N - i + 1])/2 - (T[i + 1] + T[i])/2)*G[i];
      G[i] = NTU/N*w^0.8;
      (p[i] - p[i + 1]) = kf/N*w^2;
    end for;
    w = Kt*p[N + 1]/sqrt(TIT/Tnom);
    P = w*cp*TIT*(1 - (TOP/TIP)^((gamma - 1)/gamma));
  end Plant;

  model ForwardOnDesign
    Plant plant;
  equation
    plant.w = 1;
    plant.wfg = 1;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=optdaedump --allowNonStandardModelica=initialSimplified");
  end ForwardOnDesign;
  
  model BackwardOnDesign
    Plant plant;
  equation
    plant.P = plant.Pnom;
    plant.Ts = plant.Tsnom;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=optdaedump --allowNonStandardModelica=initialSimplified");
  end BackwardOnDesign;
  
  model BackwardOffDesign
    Plant plant;
    parameter Real load = 0.3;
  equation
    plant.P = plant.Pnom*load;
    plant.Ts = plant.Tsnom;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=optdaedump --allowNonStandardModelica=initialSimplified");
  end BackwardOffDesign;
  
  model BackwardOnDesignHomotopy
    Plant plant;
    Real x(start = 0);
  equation
    0 = homotopy(x^3 + x, x); // Dummy equation to include homotopy until bug is fixed
    if initialSimplified() then
      plant.w = 1;
      plant.wfg = 1;
    else
      plant.P = homotopy(plant.Pnom, plant.Pnom);
      plant.Ts = plant.Tsnom;
    end if;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=optdaedump --allowNonStandardModelica=initialSimplified");
  end BackwardOnDesignHomotopy;
  
  model BackwardOffDesignHomotopy
    Plant plant;
    parameter Real load = 0.3;
    Real x(start = 0);
  equation
    0 = homotopy(x^3 + x, x); // Dummy equation to include homotopy until bug is fixed
    if initialSimplified() then
      plant.w = 1;
      plant.wfg = 1;
    else
      plant.P = homotopy(plant.Pnom*load, plant.Pnom);
      plant.Ts = plant.Tsnom;
    end if;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=optdaedump --allowNonStandardModelica=initialSimplified");
  end BackwardOffDesignHomotopy;
end TestInversePlant;
```

**Comment by perost** [comment]
> Here is the updated MWE

The first equation looks wrong, I assume you meant to call `initialLambda0`? You're also not using `homotopyExtended` anywhere.

**Comment by casella** [comment]
Here is another MWE to check that the experimental operators are handled correctly by other tools in the meantime

```modelica
model MWE2
  Real x(start = 0);
  Real y(start = 0);
  function homotopyExtended "Extended experimental homotopy function"
    input Real actual "Actual expression used for the simulation phase";
    input Real simplified "Simplified expression for homotopy at lambda = 0";
    input Real simplified_guess = simplified "Simplified expression used to compute the guess values of the lambda = 0 problem";
    output Real expr;
  algorithm
    expr := homotopy(actual, simplified);
  annotation(Inline = true, __OpenModelica_builtin=true);
  end homotopyExtended;
equation
  // expected solution: x = 100, y = 100
  0 = homotopyExtended((x-100)*(x + 100)*x, x - 100);
  0 = homotopyExtended((y-100)*(y + 100)*x, y - 100, y + 100);
end MWE2;
```
Tested successfully with Dymola, it requires setting `Advanced.OnlyUseHomotopyMethod = true`, otherwise homotopy is not used and the solution find with start = 0 on the actual expression is found instead.

**Comment by casella** [comment]
@phannebohm, I discussed the application of this improvement with @CJLuiten and @matteodepascali extensively.

At the end of the day, what I realized is that what is really needed is the ability to flip the causality between lambda = 0 and lambda > 0. In order to do this, you always need some kind of conditional equation, which will involve `initialLambda0()` in the switching condition. Once we need to use that, there is no need to have the extended homotopy, you can just write something as

```modelica
if initialLambda0() then
  <initial guess equations>
else
  <equations using the standard homotopy operator>
end if;
```
So, I think we can skip the `homotopyExtended()` operator definition, at least for the time being, and just focus on providing the `initialLambda0()` operator. This shouldn't really be such a big deal.

@phannebohm do you think you could implement that by the end of next week? This would allow @CJLuiten to test the prototype of the new structure of his library, and @matteodepascali to finally initialize the SOS-CO2 power plant model in backwards mode realiably, using OpenModelica.

**Comment by casella** [comment]
@phannebohm is working on this, possibly found some issues in the current implementation of homotopy, he's looking into that now.

**Comment by casella** [comment]
@phannebohm sorry to pile up more stuff on your already busy agenda. Do you have any news on this point? It would be very useful for our projects and I would also like to push it forth within MAP-Lang, but I really need to have a test implementation.

**Comment by casella** [comment]
Here is the updated MWE based on all the previous discussion, in particular about the fact that it is only necessary to introduce the `initialLambda0()` operator. I changed the name to initialSimplified(), which sounds better to me.

```modelica
model MWE "Simple test case"
  parameter Real u_fix0 = 1;
  parameter Real y_fix0 = 1;
  parameter Real y_fix = 10;
  Real u, y1, y2;

  function initialSimplified "Returns true when solving the initialization problem with the simplified expressions of homotopy() operators"
    output Boolean isInitialSimplified = false;
  annotation(Inline = true, __OpenModelica_builtin=true);
  end initialSimplified;

equation
  if initialSimplified() then
    u = u_fix0;
  else
    y2 = homotopy(y_fix, y_fix0);
  end if;
  y1 = u^3 + sin(u);
  y2 = sqrt(y1) + sin(1e-6*u);
end MWE;
```

**Comment by phannebohm** [comment]
I will finish this right after the workshop.

**Comment by casella** [comment]
BTW, I guess we should better call it `initialSimplified()`. The final decision will be up to MAP-Lang, but we better start with the right name 😅. I updated the last MWEs accordingly.

One important point. The meaning of this operator when global homotopy is performed is very clear: it returns true when solving the simplified problem at the beginning of the homotopy transformation. The question is: what if there are no `homotopy()` operators in the model? We have two options

1. `initialSimplified()` always returns false. 
2. the system is first solved with `initialSimplified() = true` and then solved again with `initialSimplified() = false`, using the previous result as initial guess

I would start by implementing option 1., then we can discuss this with the MAP-Lang group.

**Comment by casella** [comment]
Here is a test model that clearly demonstrates the benefits of the proposed approach

```modelica
package TestInversePlant
  function initialSimplified "Returns true when solving the initialization problem with the simplified expressions of homotopy() operators"
    output Boolean isInitialSimplified = false;
  annotation(Inline = true, __OpenModelica_builtin=true);
  end initialSimplified;

  model Plant
    connector RealInput = input Real;
    connector RealOutput = output Real;
  
    parameter Integer N = 5 "Number of finite volumes";
    parameter Real NTU = 4 "Nominal NTU";
    final parameter Real Gtot = NTU "Nominal total heat conductance";
    parameter Real cp = 1 "Specific heat capacity";
    parameter Real kf = 0.1 "Fluid friction coefficient";
    parameter Real Kt = 1 "Turbine flow coefficient";
    parameter Real Tnom = 700 "Nominal TIT";
    parameter Real Tsnom = 400 "Nominal flue gas outlet temperature";
    parameter Real Pnom = 223 "Nominal turbine power";
    parameter Real pnom = 1.1 "Nominal fluid inlet pressure";
    parameter Real TOP = 0.1 "Nominal turbine outlet pressure";
    parameter Real gamma = 1.2 "cp/cv";
  
    RealInput w "Fluid flow rate";
    RealInput wfg "Flue gas flow rate";
  
    Real T[N + 1] "Fluid temperatures";
    Real p[N + 1] "Fluid pressures";
    Real Tfg[N + 1] "Flue gas temperatures";
    Real G[N] "Volume conductivity";
    Real Q[N] "Thermal powers";
    Real TIT = T[N + 1] "Turbine inlet temperature";
  
    RealOutput P "Turbine power";
    RealOutput Ts = Tfg[N+1] "Outlet flue gas temperature";
    RealOutput TIP = p[N + 1] "Turbine inlet pressure";
    RealOutput pin = p[1] "Fluid inlet pressure";
  equation
    T[1] = 300;
    Tfg[1] = 800;
    for i in 1:N loop
      0 = w*cp*(T[i] - T[i + 1]) + Q[i];
      0 = wfg*cp*(Tfg[N - i + 1] - Tfg[N - i + 2]) - Q[i];
      Q[i] = ((Tfg[N - i + 2] + Tfg[N - i + 1])/2 - (T[i + 1] + T[i])/2)*G[i];
      G[i] = NTU/N*w^0.8;
      (p[i] - p[i + 1]) = kf/N*w^2;
    end for;
    w = Kt*p[N + 1]/sqrt(TIT/Tnom);
    P = w*cp*TIT*(1 - (TOP/TIP)^((gamma - 1)/gamma));
  end Plant;

  model ForwardOnDesign
    Plant plant;
  equation
    plant.w = 1;
    plant.wfg = 1;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end ForwardOnDesign;
  
  model BackwardOnDesign
    Plant plant;
  equation
    plant.P = plant.Pnom;
    plant.Ts = plant.Tsnom;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOnDesign;
  
  model BackwardOffDesign
    Plant plant;
    parameter Real load = 0.3;
  equation
    plant.P = plant.Pnom*load;
    plant.Ts = plant.Tsnom;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOffDesign;
  
  model BackwardOnDesignHomotopy
    Plant plant;
    Real x(start = 0);
  equation
    0 = homotopy(x^3 + x, x); // Dummy equation to include homotopy until bug is fixed
    if initialSimplified() then
      plant.w = 1;
      plant.wfg = 1;
    else
      plant.P = homotopy(plant.Pnom, plant.Pnom);
      plant.Ts = plant.Tsnom;
    end if;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOnDesignHomotopy;
  
  model BackwardOffDesignHomotopy
    Plant plant;
    parameter Real load = 0.3;
    Real x(start = 0);
  equation
    0 = homotopy(x^3 + x, x); // Dummy equation to include homotopy until bug is fixed
    if initialSimplified() then
      plant.w = 1;
      plant.wfg = 1;
    else
      plant.P = homotopy(plant.Pnom*load, plant.Pnom);
      plant.Ts = plant.Tsnom;
    end if;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOffDesignHomotopy;
end TestInversePlant;
```
The plant model is a toy model of a thermal power generation system: a working fluid with flow rate `w` is fed to a counter-current heat exchanger heated by a flow rate `wfg` of hot flue gas; the fluid is then expanded in a turbine, producing a power `P` and discharging the flue gas at the stack at temperature `Ts`.

If the design values of the two inputs `w` and `wfg` are prescribed, as in the `Forward` model, all the flow rates are known and most of the system equations can be solved explicitly. There is only one implicit system involving the fluid and flue gas temperatures, but that is linear, so it's no big deal to solve it. Note that _it is not necessary to provide any start value_ to solve the system.

If one instead prescribes the design values for the corresponding outputs `P` and `Ts`, as in `BackwardOnDesign` and `BackwardOffDesign`, the equations of the system become fully coupled. If appropriate start values are not provided, even a very smart nonlinear solver is very likely to fail. I tested both Dymola and OpenModelica and they both failed miserably.

The `BackwardOnDesignHomotopy` and `BackwardOffDesignHomotopy` models demonstrate the benefits of the proposed method: first, the simplified set of equations is solved, which is set up in forward mode, so that it converges without problems. Then, the found solution is used as an initial guess for the homotopy process. It is thus possible to solve the (on- and off-design) inverse problems in a reliable and robust way, without any need to worry about start values, by leveraging on the solution of the simpler direct on-design problem.

**Comment by phannebohm** [comment]
> BTW, I guess we should better call it `initialSimplified()`. The final decision will be up to MAP-Lang, but we better start with the right name 😅. I updated the last MWEs accordingly.

Of course, we can rename it easily later. Another option would be `initialHomotopy()` to tie it more closely to the `homotopy()` operator. You can even think about crazy things like `initialHomotopy(0.5)` being `true` when `lambda=0.5` and then having several initial systems solved one after the other... but unless there is an actual use for that I guess this is overkill.

> One important point. The meaning of this operator when global homotopy is performed is very clear: it returns true when solving the simplified problem at the beginning of the homotopy transformation. The question is: what if there are no `homotopy()` operators in the model? We have two options
> 
>     1. `initialSimplified()` always returns false.
> 
>     2. the system is first solved with `initialSimplified() = true` and then solved again with `initialSimplified() = false`, using the previous result as initial guess
> 
> 
> I would start by implementing option 1., then we can discuss this with the MAP-Lang group.

I think for the backend both options should be equally easy to implement. Currently we look for `homotopy()` in the initial system and only if it contains it do we even generate the lambda0 system. We could look for `initialSimplified()` at the same time. The C runtime might not work out of the box because we need to run the lambda0 system and then directly the lambda1 system without any homotopy in between. Not sure how complicated that would be to fix, perhaps it's not a big deal.

Your model `BackwardOnDesignHomotopy` doesn't really need the usual `homotopy()` operator so option 2. would be better right?

**Comment by phannebohm** [comment]
> ```modelica
> model BackwardOffDesign
>   Plant plant;
>   parameter Real load = 0.3;
> equation
>   plant.P = plant.Pnom*load;
>   plant.Tfgout = plant.Tfgoutnom;
> end BackwardOffDesign;
> ```

I guess this is a typo and the second equation should read

```modelica
plant.Ts = plant.Tsnom;
```

**Comment by casella** [comment]
> Of course, we can rename it easily later. Another option would be `initialHomotopy()` to tie it more closely to the `homotopy()` operator.

Good point, but it's only when homotopy returns simplified. `initialHomotopySimplified()` would be more accurate, but maybe too long.

I remember the discussions we had to come up with `inStream()` and `actualStream()`. They are still somewhere in old Modelica Design Meeting minutes, took us quite some time to settle them down 😄 

> You can even think about crazy things like `initialHomotopy(0.5)` being `true` when `lambda=0.5` and then having several initial systems solved one after the other... but unless there is an actual use for that I guess this is overkill.

I guess so. Also much harder to swallow for MAP-Lang.

**Comment by casella** [comment]
> Your model `BackwardOnDesignHomotopy` doesn't really need the usual `homotopy()` operator so option 2. would be better right?

Yes, but that's a toy model. In real life, homotopy will likely be used somewhere else for initialization, so it doesn't really matter. I'm really not sure what is the best choice. For the time being, please pick the easiest one to implement, then we'll see what MAP-Lang comes up with.

**Comment by phannebohm** [comment]
> I remember the discussions we had to come up with `inStream()` and `actualStream()`. They are still somewhere in old Modelica Design Meeting minutes, took us quite some time to settle them down 😄

I don't remember who said it first, but naming things is one of the hardest problems. Maybe you noticed we had the same problem with Karim's talk on implicit algorithms and the possible matching outcomes. How do you call an output that is actually matched as an input, and so on...

It feels silly but it's important to get it right so people can work with it.

**Comment by cbu-hw** [comment]
> I don't remember who said it first, but naming things is one of the hardest problems.

Phil Karlton, apparently: https://martinfowler.com/bliki/TwoHardThings.html
I personally like the first variant (with off-by-one errors) more, for its slightly tongue-in-cheek nature 😜 

> It feels silly but it's important to get it right so people can work with it.

Not at all silly! Very important, and the effort invested multiply pays off in clarity down the line (IME).

**Comment by phannebohm** [comment]
I merged #14976 so you can now use `initialSimplified` with the NB like in this test:
https://github.com/OpenModelica/OpenModelica/blob/7f884381457ed0157e8dc356fc73ea803a5b52b6/testsuite/simulation/modelica/NBackend/initialization/experimental_initialSimplified.mos#L6-L31

Unfortunately there is another bug in how the runtime handles homotopy. Apparently, it only starts the homotopy method when there is an algebraic loop that contains the homotopy operator. And it only starts initialization with the lambda0 system if it uses the homotopy method. I will try to straighten that out as soon as I have time for this.

In the meantime you should be able to work around this by including a small dummy algebraic loop that survives the backend optimizations. Something like this should be enough and has minimal overhead:
```modelica
  Real x(start = 0);
equation
  0 = homotopy(x^3 + x, x);
```

**Comment by casella** [comment]
> Unfortunately there is another bug in how the runtime handles homotopy. Apparently, it only starts the homotopy method when there is an algebraic loop that contains the homotopy operator. And it only starts initialization with the lambda0 system if it uses the homotopy method. I will try to straighten that out as soon as I have time for this.

@phannebohm don't worry about that:
- this is not really a bug, as long as we give a strict interpretation to `simplifiedInitial()`, i.e., something that deals with solving the system at $\lambda = 0$. Of course if the `homotopy()` operator is not used in the system, there is no such system;
- in practice, in all the non-trivial cases when you need to use `simplifiedInitial()`, there will always be some instance of `homotopy()` somewhere in the model.

Save your time for higher priority stuff 😃 

I'm going to close this, @matteodepascali will do some testing on tough cases (with lots of homotopy all over the place) and will open separate tickets in case there's some real issues.

Thanks!

**Comment by casella** [comment]
Oh, wait, this is only implemented in the NB, right?

Then, @matteodepascali will have to wait until all the remaining issues with the SOFC model with the NB are solved. Hope this doesn't take too much time 😅.

Would it be complicated to port this to the OB?

**Comment by phannebohm** [comment]
> Oh, wait, this is only implemented in the NB, right?

Correct.

> Would it be complicated to port this to the OB?

I think so, but I did not check.

**Comment by casella** [comment]
> I think so, but I did not check.

Never mind, let's hope the NB becomes good enough for this anytime soon.

**Comment by casella** [comment]
@phannebohm, we did some testing of my [test model](https://github.com/OpenModelica/OpenModelica/issues/11272#issuecomment-3824186901) with @matteodepascali, there are some issues.

1. If I run `TestInversePlant.ForwardOnDesign` with the old backend, the initial problem, as expected, is reported to contain 15 _linear_ equations, which are then reduced to 5 by tearing. If we run it with the new backend, the system is recognized as nonlinear (and then not torn, but that's a known issue). Why is that the case? Can't the new backend handled linear systems?
2. If I run `TestInversePlant.BackwardOnDesign`, adding one extra nonlinear equation as suggested here to make sure the whole she-bang kicks in, a very weird behaviour ensues:
  1. The `initial-lambda0` system has no implicit equations at all (except the one added for the above-mentioned reason), whereas it is expected to have the same equations as the initial section of ForwardOnDesign, i.e., 15 coupled linear equations
  2. The `initial` system is not causalized correctly: there is a non-linear system with 25 unknowns that computes all the core variables of the system, but one of them depends on `plant.P` and the equation that assigns `plant.P` comes _after_ that system. This makes no sense at all.

@phannebohm, can you please double-check? Thanks!

**Comment by phannebohm** [comment]
> Can't the new backend handled linear systems?

In the backend the system is in fact identified as linear. However, when converting to simcode we currently only generate `NONLINEAR_SYSTEM` components for algebraic loops. Not sure why we avoid `LINEAR_SYSTEM`, maybe converting to old simcode is not yet fully understood. But I guess fixing this is not high priority for now.

**Comment by phannebohm** [comment]
>    2. If I run `TestInversePlant.BackwardOnDesign`, adding one extra nonlinear equation as suggested here to make sure the whole she-bang kicks in, a very weird behaviour ensues:
> 
>    3. The `initial-lambda0` system has no implicit equations at all (except the one added for the above-mentioned reason), whereas it is expected to have the same equations as the initial section of ForwardOnDesign, i.e., 15 coupled linear equations
> 
>    4. The `initial` system is not causalized correctly: there is a non-linear system with 25 unknowns that computes all the core variables of the system, but one of them depends on `plant.P` and the equation that assigns `plant.P` comes _after_ that system. This makes no sense at all.

When I compile this slightly modified model, adding `x` and a `homotopy()` call

```modelica
  model BackwardOnDesign
    Plant plant;
    Real x(start = 0);
  equation
    0 = homotopy(x^3 + x, x);
    plant.P = plant.Pnom;
    plant.Ts = plant.Tsnom;
  end BackwardOnDesign;
```

with `-d=dumpSimCode` I get the following output, which I discuss in chunks:

```
Initial Partition (Lambda = 0)
--------------------------------
  (30) Alias of 20
  (29) Alias of 19
  (28) Alias of 18
  (27) Alias of 17
  (26) Alias of 7
  (25) Alias of 6
  (24) x := 0.0
  (23) Alias of 3
  (22) Alias of 2
  (21) Alias of 1
```

The simplified initial system looks good to me. The only difference to the initial system is the removal of `homotopy()`, as expected.

```
Initial Partition
-------------------
  (20) plant.T[1] := 300.0
  (19) plant.Tfg[1] := 800.0
  (18) plant.Tfg[6] := plant.Tsnom
  (17) Nonlinear System (size = 25, homotopy = false, mixed = false, torn = true)
  --Iteration Vars:{plant.w, plant.wfg, plant.Tfg[2], plant.Tfg[3], plant.Tfg[4], plant.Tfg[5], plant.TIT, plant.TIP, plant.p[6], $FUN_1, plant.Q[1], plant.Q[2], plant.Q[3], plant.Q[4], plant.Q[5], plant.T[2], plant.T[3], plant.T[4], plant.T[5], plant.T[6], plant.G[1], plant.G[2], plant.G[3], plant.G[4], plant.G[5]}
  --(8) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = plant.cp * plant.wfg * (plant.Tfg[6 - $i1] - plant.Tfg[7 - $i1]) - plant.Q[$i1];
  --end for;
  --(9) 0 = plant.T[6] - plant.TIT
  --(10) 0 = plant.TIT * plant.cp * plant.w * (1.0 - (plant.TOP / plant.TIP) ^ ((plant.gamma - 1.0) / plant.gamma)) - plant.Pnom
  --(11) 0 = plant.p[6] - plant.TIP
  --(12) 0 = (plant.p[6] * plant.Kt) / $FUN_1 - plant.w
  --(13) 0 = sqrt(plant.TIT / plant.Tnom) - $FUN_1
  --(14) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = plant.cp * plant.w * (plant.T[$i1] - plant.T[1 + $i1]) + plant.Q[$i1];
  --end for;
  --(15) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = (0.5 * (plant.Tfg[7 - $i1] + plant.Tfg[6 - $i1]) - 0.5 * (plant.T[1 + $i1] + plant.T[$i1])) * plant.G[$i1] - plant.Q[$i1];
  --end for;
  --(16) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = (0.2 * plant.NTU) * plant.w ^ 0.8 - plant.G[$i1];
  --end for;
  (7) resizable call [index  0]
  (6) plant.pin := plant.p[1]
  (5) Nonlinear System (size = 1, homotopy = true, mixed = false, torn = true)
  --Iteration Vars:{x}
  --(4) 0 = homotopy(x ^ 3.0 + x, x)
  (3) plant.P := plant.Pnom
  (2) plant.Ts := plant.Tsnom
  (1) plant.Gtot := plant.NTU
```

The algebraic loop `(17)` does no longer depend on `plant.P` (with capital P) directly because that was replaced by the alias module with `plant.Pnom`. But `(17)` computes `plant.p[6]` (with small p) which is used in component `(7)`, a resizable call that is listed further below with `Generic Calls`. It explicitly solves `plant.p[i]` based on `plant.p[i+1]`, i.e., in reverse order. So the initial system look correct to me.

```
Algebraic Partition 1
-----------------------
  (47) Nonlinear System (size = 1, homotopy = false, mixed = false, torn = true)
  --Iteration Vars:{x}
  --(46) 0 = x ^ 3.0 + x
```

As expected `x` is put in its own partition as it is independent of the rest of the system. Note that we don't partition the initial system because it is probably not worth the effort.

```
Algebraic Partition 2
-----------------------
  (45) Alias of 19
  (44) Alias of 18
  (43) Alias of 20
  (42) Nonlinear System (size = 25, homotopy = false, mixed = false, torn = true)
  --Iteration Vars:{plant.wfg, plant.w, plant.Tfg[2], plant.Tfg[3], plant.Tfg[4], plant.Tfg[5], plant.G[1], plant.G[2], plant.G[3], plant.G[4], plant.G[5], plant.T[2], plant.T[3], plant.T[4], plant.T[5], plant.T[6], plant.TIT, plant.TIP, plant.p[6], $FUN_1, plant.Q[1], plant.Q[2], plant.Q[3], plant.Q[4], plant.Q[5]}
  --(33) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = (0.5 * (plant.Tfg[7 - $i1] + plant.Tfg[6 - $i1]) - 0.5 * (plant.T[1 + $i1] + plant.T[$i1])) * plant.G[$i1] - plant.Q[$i1];
  --end for;
  --(34) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = (0.2 * plant.NTU) * plant.w ^ 0.8 - plant.G[$i1];
  --end for;
  --(35) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = plant.cp * plant.wfg * (plant.Tfg[6 - $i1] - plant.Tfg[7 - $i1]) - plant.Q[$i1];
  --end for;
  --(36) 0 = plant.T[6] - plant.TIT
  --(37) 0 = plant.TIT * plant.cp * plant.w * (1.0 - (plant.TOP / plant.TIP) ^ ((plant.gamma - 1.0) / plant.gamma)) - plant.Pnom
  --(38) 0 = plant.p[6] - plant.TIP
  --(39) 0 = (plant.p[6] * plant.Kt) / $FUN_1 - plant.w
  --(40) 0 = sqrt(plant.TIT / plant.Tnom) - $FUN_1
  --(41) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = plant.cp * plant.w * (plant.T[$i1] - plant.T[1 + $i1]) + plant.Q[$i1];
  --end for;
  (32) Alias of 7
  (31) Alias of 6
```

The regular system is an alias of the initial system. The variables and equations in the algebraic loop are in different order so it is not detected as alias of (17), we probably should have better equality checks for strong components.

```
Event Partition
-----------------
```

Nothing to see here, there are no events.

```
Generic Calls
---------------
  (0) [SNGL]: {{$i1 | start:5, step:-1, stop:1, size: 5}}
	plant.p[$i1] = plant.p[1 + $i1] + (0.2 * plant.kf) * plant.w ^ 2.0
```

This is the call mentioned in component `(7) resizable call [index  0]`.

So everything looks correct to me. Am I missing something?

**Comment by phannebohm** [comment]
> The regular system is an alias of the initial system. The variables and equations in the algebraic loop are in different order so it is not detected as alias of (17), we probably should have better equality checks for strong components.

I fixed the alias miss with #15499, once that is merged we should see more easily what is different between the regular/initial systems.

**Comment by casella** [comment]
> > Can't the new backend handled linear systems?
> 
> In the backend the system is in fact identified as linear. 

Good. 

However, when converting to simcode we currently only generate `NONLINEAR_SYSTEM` components for algebraic loops. Not sure why we avoid `LINEAR_SYSTEM`, maybe converting to old simcode is not yet fully understood.

OK.

> But I guess fixing this is not high priority for now.

Not really. Nonlinear solvers can solve linear systems, though not in the most efficient way. This needs to be fixed once sparsity pattern handling is thoroughly done. This is now tracked by @15500.

**Comment by casella** [comment]
Indeed you are right, sorry, I shouldn't work late hours 😄 The examples without homotopy work as expected.

The actual problems are with the two test cases using homotopy and the new initialSimplified() operator. I added the dummy equation to the [MWE](https://github.com/OpenModelica/OpenModelica/issues/11272#issuecomment-3824186901), and also all the compile flags, so we're sure to be on the same page 😅 

Try compiling `TestInversePlant.BackwardOnDesignHomotopy`. The compile flags are `"--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified"`, which should be correct.

You get:

```
Initial Partition (Lambda = 0) 
-------------------------------- 
  (30) Alias of 20 
  (29) Alias of 19 
  (28) Alias of 18 
  (27) plant.P := plant.Pnom 
  (26) Alias of 16 
  (25) Alias of 6 
  (24) Alias of 5 
  (23) x := 0.0 
  (22) Alias of 2 
  (21) Alias of 1 

Initial Partition 
------------------- 
  (20) plant.T[1] := 300.0 
  (19) plant.Tfg[1] := 800.0 
  (18) plant.Tfg[6] := plant.Tsnom 
  (17) plant.P := homotopy(plant.Pnom, plant.Pnom) 
  (16) Nonlinear System (size = 25, homotopy = false, mixed = false, torn = true) 
  --Iteration Vars:{plant.w, plant.TIT, plant.T[2], plant.T[3], plant.T[4], plant.T[5], plant.T[6], plant.Q[1], plant.Q[2], plant.Q[3], ...} 
  --(7) 0 = plant.TIT * plant.cp * plant.w * (1.0 - (plant.TOP / plant.TIP) ^ ((plant.gamma - 1.0) / plant.gamma)) - plant.P 
  --(8) 0 = plant.T[6] - plant.TIT 
  --(9) For-Loop-Residual: 
  --for {$i1 in 1:5} loop 
  --  0 = (0.5 * (plant.Tfg[7 - $i1] + plant.Tfg[6 - $i1]) - 0.5 * (plant.T[1 + $i1] + plant.T[$i1])) * plant.G[$i1] - plant.Q[$i1]; 
  --end for; 
  --(10) For-Loop-Residual: 
  --for {$i1 in 1:5} loop 
  --  0 = plant.cp * plant.w * (plant.T[$i1] - plant.T[1 + $i1]) + plant.Q[$i1]; 
  --end for; 
  --(11) For-Loop-Residual: 
  --for {$i1 in 1:5} loop 
  --  0 = plant.cp * plant.wfg * (plant.Tfg[6 - $i1] - plant.Tfg[7 - $i1]) - plant.Q[$i1]; 
  --end for; 
  --(12) For-Loop-Residual: 
  --for {$i1 in 1:5} loop 
  --  0 = (0.2 * plant.NTU) * plant.w ^ 0.8 - plant.G[$i1]; 
  --end for; 
  --(13) 0 = plant.p[6] - plant.TIP 
  --(14) 0 = (plant.p[6] * plant.Kt) / $FUN_1 - plant.w 
  --(15) 0 = sqrt(plant.TIT / plant.Tnom) - $FUN_1 
  (6) resizable call [index  0] 
  (5) plant.pin := plant.p[1] 
  (4) Nonlinear System (size = 1, homotopy = true, mixed = false, torn = true) 
  --Iteration Vars:{x} 
  --(3) 0 = homotopy(x ^ 3.0 + x, x) 
  (2) plant.Ts := plant.Tsnom 
  (1) plant.Gtot := plant.NTU 
```
Here is the problem: the initial-lambda0 equations are basically solved in the same way the (homotopy-based) initial equations are. They're all aliases, except (27), which rightfully uses the simplified branch only. In particular, eq. (26) is an alias of eq. (16), which is the same nonlinear system with 25 equations that you get in the initial partition of the TestInversePlant.BackwardOnDesign problem. 

That's not good. The whole point of writing

```modelica
  if initialSimplified() then
    plant.w = 1;
    plant.wfg = 1;
  else
    plant.P = homotopy(plant.Pnom, plant.Pnom);
    plant.Ts = plant.Tsnom;
  end if;
```
is precisely that the two closure equations at $\lambda = 0$ are the same as in `TestInversePlant.ForwardOnDesign`, so the problem at $\lambda = 0$ should have 15 linear equations (possibly treated as nonlinear, but still linear), not 25 nonlinear ones.

I'm not sure why this happens, but apparently there is still something not OK in the handling of initialSimplified(). Please check.

**Comment by phannebohm** [comment]
Took me a while. The problem is your definition of `initialSimplified()`.

1. The function must be defined not within any package, otherwise you have calls of the form `TestInversePlant.initialSimplified()` which are not recognized by the backend.
2. If you violate 1. and set `annotation(Inline = true)` the frontend inlines the call before it reaches the backend. This is why I couldn't find it anywhere in the dumps. 

If you define `initialSimplified` outside the package it works and for `TestInversePlant.BackwardOnDesignHomotopy` gives

```
Initial Partition (Lambda = 0)
--------------------------------
  (39) plant.w := 1.0
  (38) plant.wfg := 1.0
  (37) Alias of 20
  (36) Alias of 19
  (35) resizable call [index  1]
  (34) Nonlinear System (size = 15, homotopy = false, mixed = false, torn = true)
  --Iteration Vars:{plant.T[2], plant.T[3], plant.T[4], plant.T[5], plant.T[6], plant.Q[1], plant.Q[2], plant.Q[3], plant.Q[4], plant.Q[5], ...}
  --(31) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = (0.5 * (plant.Tfg[7 - $i1] + plant.Tfg[6 - $i1]) - 0.5 * (plant.T[1 + $i1] + plant.T[$i1])) * plant.G[$i1] - plant.Q[$i1];
  --end for;
  --(32) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = plant.cp * plant.w * (plant.T[$i1] - plant.T[1 + $i1]) + plant.Q[$i1];
  --end for;
  --(33) For-Loop-Residual:
  --for {$i1 in 1:5} loop
  --  0 = plant.cp * plant.wfg * (plant.Tfg[6 - $i1] - plant.Tfg[7 - $i1]) - plant.Q[$i1];
  --end for;
  (30) plant.TIT := plant.T[6]
  (29) $FUN_1 := sqrt(plant.TIT / plant.Tnom)
  (28) plant.p[6] := plant.w / ((plant.Kt * $FUN_1) / $FUN_1 ^ 2.0)
  (27) plant.TIP := plant.p[6]
  (26) plant.P := plant.TIT * plant.cp * plant.w * (1.0 - (plant.TOP / plant.TIP) ^ ((plant.gamma - 1.0) / plant.gamma))
  (25) plant.Ts := plant.Tfg[6]
  (24) Alias of 5
  (23) Alias of 4
  (22) x := 0.0
  (21) Alias of 1
```

**Comment by phannebohm** [comment]
Also I completely forgot myself, but if you use `--allowNonStandardModelica=initialSimplified` then you don't need any `homotopy` calls in the model. It also works if there are only `initialSimplified` calls.

**Comment by phannebohm** [comment]
I noticed a separate issue with `TestInversePlant.BackwardOffDesignHomotopy`. When there is an algebraic loop depending on `homotopy` but not containing it itself, then the runtime does no homotopy steps. In this case the loop depends on `plant.P` but has no `homotopy` calls of its own.

This is independent of `initialSimplified` so I opened #15509 for that.

**Comment by casella** [comment]
> Took me a while. 

Sorry about that 😅 

> If you define `initialSimplified` outside the package it works

That's good news 😃 

> The problem is your definition of `initialSimplified()`.
> 
>   1. The function must be defined not within any package, otherwise you have calls of the form `TestInversePlant.initialSimplified()` which are not recognized by the backend.

As I understand, according to the [MLS Sect. 13.4](https://specification.modelica.org/maint/3.6/packages.html#file-system-mapping-of-package-class), a directory can only contain a package, and a .mo file can only contain a single class with name matching the filename. So, I understand we should have a separate file initialSimplified.mo which only contains the function definition and should be loaded separaterly. Is that correct?

>     2. If you violate 1. and set `annotation(Inline = true)` the frontend inlines the call before it reaches the backend. This is why I couldn't find it anywhere in the dumps.

OK.

I'll take care of this tomorrow, now it's too late.

**Comment by casella** [comment]
> I noticed a separate issue with `TestInversePlant.BackwardOffDesignHomotopy`. When there is an algebraic loop depending on `homotopy` but not containing it itself, then the runtime does no homotopy steps. In this case the loop depends on `plant.P` but has no `homotopy` calls of its own.

Whoa, that's bad. We need global homotopy, not local, that's what the MLS says.

> This is independent of `initialSimplified` so I opened [#15509](https://github.com/OpenModelica/OpenModelica/issues/15509) for that.

I'll comment there.

**Comment by casella** [comment]
Here's the updated MWE with initialSimplified() in the global scope, dummy equations, and all required flags. The function should be loaded from a separate file, since it is not legal to have two top-level classes in the same .mo file.
```modelica
function initialSimplified "Returns true when solving the initialization problem with the simplified expressions of homotopy() operators"
    output Boolean isInitialSimplified = false;
  annotation(Inline = true, __OpenModelica_builtin=true);
end initialSimplified;

package TestInversePlant
  model Plant
    connector RealInput = input Real;
    connector RealOutput = output Real;
  
    parameter Integer N = 5 "Number of finite volumes";
    parameter Real NTU = 4 "Nominal NTU";
    final parameter Real Gtot = NTU "Nominal total heat conductance";
    parameter Real cp = 1 "Specific heat capacity";
    parameter Real kf = 0.1 "Fluid friction coefficient";
    parameter Real Kt = 1 "Turbine flow coefficient";
    parameter Real Tnom = 700 "Nominal TIT";
    parameter Real Tsnom = 400 "Nominal flue gas outlet temperature";
    parameter Real Pnom = 223 "Nominal turbine power";
    parameter Real pnom = 1.1 "Nominal fluid inlet pressure";
    parameter Real TOP = 0.1 "Nominal turbine outlet pressure";
    parameter Real gamma = 1.2 "cp/cv";
  
    RealInput w "Fluid flow rate";
    RealInput wfg "Flue gas flow rate";
  
    Real T[N + 1] "Fluid temperatures";
    Real p[N + 1] "Fluid pressures";
    Real Tfg[N + 1] "Flue gas temperatures";
    Real G[N] "Volume conductivity";
    Real Q[N] "Thermal powers";
    Real TIT = T[N + 1] "Turbine inlet temperature";
  
    RealOutput P "Turbine power";
    RealOutput Ts = Tfg[N+1] "Outlet flue gas temperature";
    RealOutput TIP = p[N + 1] "Turbine inlet pressure";
    RealOutput pin = p[1] "Fluid inlet pressure";
  equation
    T[1] = 300;
    Tfg[1] = 800;
    for i in 1:N loop
      0 = w*cp*(T[i] - T[i + 1]) + Q[i];
      0 = wfg*cp*(Tfg[N - i + 1] - Tfg[N - i + 2]) - Q[i];
      Q[i] = ((Tfg[N - i + 2] + Tfg[N - i + 1])/2 - (T[i + 1] + T[i])/2)*G[i];
      G[i] = NTU/N*w^0.8;
      (p[i] - p[i + 1]) = kf/N*w^2;
    end for;
    w = Kt*p[N + 1]/sqrt(TIT/Tnom);
    P = w*cp*TIT*(1 - (TOP/TIP)^((gamma - 1)/gamma));
  end Plant;

  model ForwardOnDesign
    Plant plant;
  equation
    plant.w = 1;
    plant.wfg = 1;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end ForwardOnDesign;
  
  model BackwardOnDesign
    Plant plant;
  equation
    plant.P = plant.Pnom;
    plant.Ts = plant.Tsnom;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOnDesign;
  
  model BackwardOffDesign
    Plant plant;
    parameter Real load = 0.3;
  equation
    plant.P = plant.Pnom*load;
    plant.Ts = plant.Tsnom;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOffDesign;
  
  model BackwardOnDesignHomotopy
    Plant plant;
    Real x(start = 0);
  equation
    0 = homotopy(x^3 + x, x); // Dummy equation to include homotopy until bug is fixed
    if initialSimplified() then
      plant.w = 1;
      plant.wfg = 1;
    else
      plant.P = homotopy(plant.Pnom, plant.Pnom);
      plant.Ts = plant.Tsnom;
    end if;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOnDesignHomotopy;
  
  model BackwardOffDesignHomotopy
    Plant plant;
    parameter Real load = 0.3;
    Real x(start = 0);
  equation
    0 = homotopy(x^3 + x, x); // Dummy equation to include homotopy until bug is fixed
    if initialSimplified() then
      plant.w = 1;
      plant.wfg = 1;
    else
      plant.P = homotopy(plant.Pnom*load, plant.Pnom);
      plant.Ts = plant.Tsnom;
    end if;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOffDesignHomotopy;
end TestInversePlant;
```

**Comment by casella** [comment]
New version of the MWE without the dummy equations, they should be no longer needed. 

IMPORTANT: the initialSimplified() function must be loaded in the global scope.

```modelica
impure function initialSimplified "Returns true when solving the initialization problem with the simplified expressions of homotopy() operators"
    output Boolean isInitialSimplified = false;
  annotation(Inline = true, __OpenModelica_builtin=true);
end initialSimplified;

package TestInversePlant
  model Plant
    connector RealInput = input Real;
    connector RealOutput = output Real;
  
    parameter Integer N = 5 "Number of finite volumes";
    parameter Real NTU = 4 "Nominal NTU";
    final parameter Real Gtot = NTU "Nominal total heat conductance";
    parameter Real cp = 1 "Specific heat capacity";
    parameter Real kf = 0.1 "Fluid friction coefficient";
    parameter Real Kt = 1 "Turbine flow coefficient";
    parameter Real Tnom = 700 "Nominal TIT";
    parameter Real Tsnom = 400 "Nominal flue gas outlet temperature";
    parameter Real Pnom = 223 "Nominal turbine power";
    parameter Real pnom = 1.1 "Nominal fluid inlet pressure";
    parameter Real TOP = 0.1 "Nominal turbine outlet pressure";
    parameter Real gamma = 1.2 "cp/cv";
  
    RealInput w "Fluid flow rate";
    RealInput wfg "Flue gas flow rate";
  
    Real T[N + 1] "Fluid temperatures";
    Real p[N + 1] "Fluid pressures";
    Real Tfg[N + 1] "Flue gas temperatures";
    Real G[N] "Volume conductivity";
    Real Q[N] "Thermal powers";
    Real TIT = T[N + 1] "Turbine inlet temperature";
  
    RealOutput P "Turbine power";
    RealOutput Ts = Tfg[N+1] "Outlet flue gas temperature";
    RealOutput TIP = p[N + 1] "Turbine inlet pressure";
    RealOutput pin = p[1] "Fluid inlet pressure";
  equation
    T[1] = 300;
    Tfg[1] = 800;
    for i in 1:N loop
      0 = w*cp*(T[i] - T[i + 1]) + Q[i];
      0 = wfg*cp*(Tfg[N - i + 1] - Tfg[N - i + 2]) - Q[i];
      Q[i] = ((Tfg[N - i + 2] + Tfg[N - i + 1])/2 - (T[i + 1] + T[i])/2)*G[i];
      G[i] = NTU/N*w^0.8;
      (p[i] - p[i + 1]) = kf/N*w^2;
    end for;
    w = Kt*p[N + 1]/sqrt(TIT/Tnom);
    P = w*cp*TIT*(1 - (TOP/TIP)^((gamma - 1)/gamma));
  end Plant;

  model ForwardOnDesign
    Plant plant;
  equation
    plant.w = 1;
    plant.wfg = 1;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end ForwardOnDesign;
  
  model BackwardOnDesign
    Plant plant;
  equation
    plant.P = plant.Pnom;
    plant.Ts = plant.Tsnom;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOnDesign;
  
  model BackwardOffDesign
    Plant plant;
    parameter Real load = 0.3;
  equation
    plant.P = plant.Pnom*load;
    plant.Ts = plant.Tsnom;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOffDesign;
  
  model BackwardOnDesignHomotopy
    Plant plant;
  equation
    if initialSimplified() then
      plant.w = 1;
      plant.wfg = 1;
    else
      plant.P = homotopy(plant.Pnom, plant.Pnom);
      plant.Ts = plant.Tsnom;
    end if;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOnDesignHomotopy;
  
  model BackwardOffDesignHomotopy
    Plant plant;
    parameter Real load = 0.3;
  equation
    if initialSimplified() then
      plant.w = 1;
      plant.wfg = 1;
    else
      plant.P = homotopy(plant.Pnom*load, plant.Pnom);
      plant.Ts = plant.Tsnom;
    end if;
  annotation(__OpenModelica_commandLineOptions = "--newBackend -d=dumpSimCode --allowNonStandardModelica=initialSimplified");
  end BackwardOffDesignHomotopy;
end TestInversePlant;
```

**Comment by casella** [comment]
@perost if I load the package and the function shown in the previous post and then flatten TestInversePlant.BackwardOffDesignHomotopy, the flat model has no trace of `initialSimplified()`. In a previous [comment](https://github.com/OpenModelica/OpenModelica/issues/11272#issuecomment-3357353617) of yours you wrote:

> But if the name doesn't matter then I guess you can already achieve what you want by using the __OpenModelica_builtin=true annotation, ***which will make the frontend assume it's a builtin function and just pass it through***.

Why isn't it passed through, and instead it is apparently inlined and constant evaluated to false? Am I missing something?

**Comment by perost** [comment]
> [@perost](https://github.com/perost) if I load the package and the function shown in the previous post and then flatten TestInversePlant.BackwardOffDesignHomotopy, the flat model has no trace of `initialSimplified()`. In a previous [comment](https://github.com/OpenModelica/OpenModelica/issues/11272#issuecomment-3357353617) of yours you wrote:
> 
> > But if the name doesn't matter then I guess you can already achieve what you want by using the __OpenModelica_builtin=true annotation, _**which will make the frontend assume it's a builtin function and just pass it through**_.
> 
> Why isn't it passed through, and instead it is apparently inlined and constant evaluated to false? Am I missing something?

Being builtin doesn't mean the frontend won't evaluate it. My previous advice was for the definition in https://github.com/OpenModelica/OpenModelica/issues/11272#issuecomment-3357044834, but since the definition you're using now just returns a constant value the frontend will evaluate it if you use it as an if-condition.

One solution is to declare the function as `impure`. I don't know if that would cause any issues, but that's how we declare the normal `initial`:
https://github.com/OpenModelica/OpenModelica/blob/7b0520e1344489897183d75bbe8a505b9927080a/OMCompiler/Compiler/NFFrontEnd/NFModelicaBuiltin.mo#L68-L74

**Comment by casella** [comment]
Thanks @perost! Indeed, adding impure solved the problem. BTW, @phannebohm has been doing that in _his own_ tests since the beginning 😅, that explains why he didn't have any problem.

**Comment by casella** [comment]
I'm closing this ticket because everything is done, on the NB. We will then start the MCP process later on, and possibly port it to the OB, if it is accepted by MAP-Lang.
